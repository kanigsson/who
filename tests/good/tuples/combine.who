type unit
logic tt : unit
type region ['u]
type refty ['reg 'ty]
logic ref_get ['reg 'u] : 'reg region -> ('reg,'u) refty -> 'u
logic combine [||'e1 'e2 'e3] : <'e1 'e2> -> <'e2 'e3> -> <'e1 'e2 'e3>

logic get_2_1_tuple ['a 'b] : 'a * 'b -> 'a
logic get_2_2_tuple ['a 'b] : 'a * 'b -> 'b
logic mk_2tuple ['a 'b] : 'a -> 'b -> 'a * 'b


logic = ['a] : 'a -> 'a -> prop
logic -> : prop -> prop -> prop
logic !! ['a|r|'e] : ref(r,'a) -> <r 'e > -> 'a

logic z : int
letregion a b
goal g :
  forall (u : <a >) (v : <b>) (x : ref(a,int)) (k : ref(b,int)).
    (let f = (fun (c : <a b>) -> !!x@c = !!k@c) in
    f (combine[{a} {} {b}] u v))
