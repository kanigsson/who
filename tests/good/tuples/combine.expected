type unit
logic tt  : unit
type region['u||]
type refty['reg 'ty||]
logic ref_get ['reg 'u||] : region[reg] -> refty[reg,u] -> u
logic get_2_1_tuple ['a 'b||] : (a * b) -> a
logic get_2_2_tuple ['a 'b||] : (a * b) -> b
logic mk_2tuple ['a 'b||] : a -> b -> a * b
logic = ['a||] : a -> a -> prop
logic ->  : prop -> prop -> prop
logic z  : int
INTROS ['b 'a||]
goal g:
  forall (u:region[a]).
  forall (v:region[b]).
  forall (x:refty[a,int]).
  forall (k:refty[b,int]).
  let
  f =
     (fun (c:region[b] * region[a]) ->
     (ref_get [a,int||] (get_2_2_tuple [region[b],region[a]||] c) x) =
     (ref_get [b,int||] (get_2_1_tuple [region[b],region[a]||] c) k))
  in f (mk_2tuple [region[b],region[a]||] v u)
